presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java

```diff
@@ -0,0 +1,159 @@
+import com.google.common.util.concurrent.ListeningExecutorService;
+import io.airlift.concurrent.MoreFutures;
+import io.airlift.slice.InputStreamSliceInput;
+import io.airlift.slice.OutputStreamSliceOutput;
+import io.airlift.slice.SliceOutput;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class BinaryFileSingleStreamSpiller
+        implements SingleStreamSpiller
+{
+    private final Path targetFile;
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final SliceOutput output;
+    private final ListeningExecutorService executor;
+    private final Queue<CompletableFuture<?>> scheduledSpills = new ArrayDeque<>();
+
+    private boolean closed;
+
+    public BinaryFileSingleStreamSpiller(BlockEncodingSerde blockEncodingSerde, ListeningExecutorService executor, Path spillPath)
+    {
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        try {
+            targetFile = Files.createTempFile(spillPath, "presto-spill", ".bin");
+            output = new OutputStreamSliceOutput(new BufferedOutputStream(new FileOutputStream(targetFile.toFile())));
+        }
+        catch (IOException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to create spill directory", e);
+        }
+    }
+
+    @Override
+    public synchronized CompletableFuture<?> spill(Iterator<Page> pageIterator)
+    {
+        CompletableFuture<?> future = MoreFutures.toCompletableFuture(executor.submit(
+            () -> writePages(pageIterator)
+        ));
+        scheduledSpills.add(future);
+        removeFinishedFutures();
+        return future;
+    }
+
+    @Override
+    public synchronized CompletableFuture<?> spill(Page page)
+    {
+        CompletableFuture<?> future = MoreFutures.toCompletableFuture(executor.submit(
+                () -> writePage(page)
+        ));
+        scheduledSpills.add(future);
+        removeFinishedFutures();
+        return future;
+    }
+
+    private void writePage(Page page)
+    {
+        synchronized (output) {
+            PagesSerde.writePages(blockEncodingSerde, output, page);
+        }
+    }
+
+    private void writePages(Iterator<Page> pageIterator)
+    {
+        synchronized (output) {
```

Comment:
it's not a bug. Those two methods are executed in executor and are long blocking operations (writes do disk). Added some comments